### PR TITLE
saas6 caching/reqs improvements

### PR DESCRIPTION
**Pull request by xmo-odoo**
Ignore  the SQL tracing commit at the top for review.

Tries to avoid extra queries by not doing unnecessary work, reusing probably existing env caches and attempting to add new ormcaches.

Can probably be cherrypicked if only some of them are considered of interest.

As of Friday 27th, worked mostly on small potato pages (index & contact form), trying out bigger tickets (100+ queries _after_ cache warmups)

(funny story: apparently, each new post adds 3 queries to the forums home)

/cc @rco-odoo @odony @KangOl

### CODE REVIEW

**Comment by odony** [comment]
:+1: for the caching of ICP, re-auth and main_menu optimizations, I'll cherry-pick them. The `show_address` optimization is much more specific and thus perhaps less useful.

**Comment by xmo-odoo** [comment]
New commit: fix bulk-calculation of voting stuff on forum posts, with warm caches saves ~2 requests per post on the forum front page.

**Comment by rco-odoo** [comment]
:+1: on the vote metadata

**Comment by rco-odoo** [comment]
@odony I don't see a way to avoid this, since `BaseModel` must be defined before model `ir.ui.view`...  However, I have a better solution with the (not yet merged) new API of `ormcache`: select all key/val pairs in context such that key ends with `'_view_ref'`

    @ormcache(..., "frozenset((k, context[k]) for k in context if k.endswith('_view_ref'))")

**Comment by rco-odoo** [comment]
@odony You are completely right, I forgot about the invalidation part.
FYI, `read_combined` is currently **not** ormcached.

**Comment by odony** [comment]
@rco-odoo Regarding the list of view types, you could just do it the other way around: define the list of types in the ORM itself (a constant somewhere) and have ir.ui.view use it.
But anyway this is only fixing a very small part of the giant invalidation problem ;-)

Regarding the caching of read_combined, that's actually a good point! We do cache `_read_template`, which also relies on `read_combined`, but not `read_combined` itself.
Why don't we do try that, it's much easier to invalidate, and that invalidation should already be done for the`read_template` invalidation?

After calling `read_combined`, `fields_view_get` probably wastes a lot of time in `postprocess_and_fields`, which another thing we should be able to optimize separately without caching. The hard part should be loading translations and checking access rights, and both are already cached separately.

To sum it up: I would find it safer and easier not to cache `fields_view_get` at all, but make sure we optimize the methods it relies on, each of them relying on their own cached data, and already invalidated separately.

**Comment by rco-odoo** [comment]
@odony I made a quick test where I decorate the method `get_actions` of `ir.values` with `ormcache`, and the result is impressive: the call to `fields_view_get` made by the Customer list issues 12 queries instead of 62 (originally 73)!

The method depends on models `ir.values` and `ir.actions.*`. I will make any modification to one of those models invalidate the method cache.

File: addons/portal/mail_message.py [hunk 1]
```diff
@@ -19,45 +19,17 @@
...
-                        (self._description, operation))
-
-        return super(mail_message, self).check_access_rule(cr, uid, ids=ids, operation=operation, context=context)
+    def _compute_domain(self, cr, uid, model_name, mode="read"):
+        domain = super(IrRule, self)._compute_domain(cr, uid, model_name, mode=mode)
+        if model_name == 'mail.message' and self.pool['res.users'].has_group(cr, uid, 'base.group_user'):
+            domain = expression.AND([
+                domain,
+                ['&', ('type', '=', 'comment'), ('subtype_id', '=', False)]
+            ])
```
Comment (rco-odoo) [inline addons/portal/mail_message.py]: @xmo-odoo I think the condition is incorrect. The domain returned by `_compute_domain()` defines which records you have access to, and you cannot access records that satisfy `type='comment' AND subtype_id IS NULL`. Therefore you have to negate this condition in `_compute_domain()`:

    domain = expression.AND([domain, [`|`, ('type', '!=', 'comment'), ('subtype_id', '!=', False)]])

or maybe something simpler, like what `_search()` does:

    domain = expression.AND([domain, [('subtype_id', '!=', False)]])

File: addons/website_mail/models/mail_message.py [hunk 2]
```diff
@@ -60,30 +57,12 @@ def default_get(self, cr, uid, fields_list, context=None):
...
-            group_user_id = self.pool.get("ir.model.data").get_object_reference(cr, uid, 'base', 'group_public')[1]
-            if group_user_id in [group.id for group in group_ids]:
-                cr.execute('SELECT id FROM "%s" WHERE website_published IS FALSE AND id = ANY (%%s)' % (self._table), (ids,))
-                if cr.fetchall():
-                    raise AccessError(_('The requested operation cannot be completed due to security restrictions. Please contact your system administrator.\n\n(Document type: %s, Operation: %s)') % (self._description, operation))
-        return super(MailMessage, self).check_access_rule(cr, uid, ids=ids, operation=operation, context=context)
+    def _compute_domain(self, cr, uid, model_name, mode="read"):
+        domain = super(IrRule, self)._compute_domain(cr, uid, model_name, mode=mode)
+        if model_name == 'mail.message' and self.pool['res.users'].has_group(cr, uid, 'base.group_public'):
+            domain = expression.AND([domain, [('website_published', '=', True)]])
```
Comment (rco-odoo) [inline addons/website_mail/models/mail_message.py]: @xmo-odoo This one is correct.

File: addons/portal/mail_message.py [hunk 3]
```diff
@@ -19,45 +19,17 @@
...
-        if group_user_id not in [group.id for group in group_ids]:
-            cr.execute('SELECT DISTINCT id FROM "%s" WHERE type = %%s AND subtype_id IS NULL AND id = ANY (%%s)' % (self._table), ('comment', ids,))
-            if cr.fetchall():
-                raise AccessError(_('The requested operation cannot be completed due to security restrictions. Please contact your system administrator.\n\n(Document type: %s, Operation: %s)') % \
-                        (self._description, operation))
-
-        return super(mail_message, self).check_access_rule(cr, uid, ids=ids, operation=operation, context=context)
+    def _compute_domain(self, cr, uid, model_name, mode="read"):
+        domain = super(IrRule, self)._compute_domain(cr, uid, model_name, mode=mode)
+        if model_name == 'mail.message' and self.pool['res.users'].has_group(cr, uid, 'base.group_user'):
```
Comment (KangOl) [inline addons/portal/mail_message.py]: Should be `not has_group`

**Comment by rco-odoo** [comment]
The commit 711043d was been cherry-picked and included in #6600.
